proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -368,6 +368,1382 @@ lemma invoke_sched_context_corres:
      reused in Finalise and IpcCancel *)
   sorry
 
+(* FIXME RT: move *)
+lemmas update_sched_context_decompose_ext
+  = update_sched_context_decompose[where f="f x" and g="g y" for f g x y]
+lemmas update_sched_context_decompose2
+  = update_sched_context_decompose[where g="\<lambda>sc. g (h sc)" for g h]
+lemmas update_sched_context_decompose_ext2
+  = update_sched_context_decompose2[where f="f x" and g="g y" for f g x y]
+
+lemma refill_new_bundled:
+   "refill_new sc_ptr max_refills budget period
+    = (do cur_time \<leftarrow> gets cur_time;
+          set_sc_obj_ref sc_period_update sc_ptr period;
+          set_sc_obj_ref sc_budget_update sc_ptr budget;
+          update_sched_context sc_ptr (\<lambda>sc. sc\<lparr>sc_refill_max := max_refills,
+                                               sc_refills := [\<lparr>r_time = cur_time, r_amount = budget\<rparr>]\<rparr>);
+          maybe_add_empty_tail sc_ptr
+       od)"
+  unfolding refill_new_def
+            update_sched_context_decompose_ext2[where f=sc_refills_update and g=sc_refill_max_update]
+            update_sched_context_decompose_ext2[where f=sc_refill_max_update and g=sc_budget_update]
+            update_sched_context_decompose_ext2[where f=sc_budget_update and g=sc_period_update]
+  apply (simp add: bind_assoc)
+  done
+
+(* FIXME RT: move *)
+lemmas sc_inv_state_eq' = getObject_sc_inv[THEN use_valid[rotated], rotated
+                                           , where s=s and P="(=) s" for s, OF _ refl]
+
+lemma sc_inv_state_eq:
+  "(a :: sched_context, s') \<in> fst (getSchedContext p s) \<Longrightarrow> s' = s"
+  apply (fastforce dest: sc_inv_state_eq' simp: getSchedContext_def)
+  done
+
+lemma no_ofail_readSchedContext:
+  "no_ofail (sc_at' p) (readSchedContext p)"
+  unfolding readSchedContext_def by wpsimp
```

Comment:
This one and the rewrite lemmas that follow have better places to go... maybe we can move them all to KHeap_R.